lib/evaluator.py

```diff
@@ -50,25 +51,79 @@ def calculate_recall(self, ratings, predictions):
     def recall_at_x(self, n_recommendations, predictions):
         """
         The method calculates the average recall of all users by only looking at the top n_recommendations
+        and the normalized Discounted Cumulative Gain.
 
         :param int n_recommendations: number of recommendations to look at, sorted by relevance.
         :param float[][] predictions: calculated predictions of the recommender
-        :returns: Recall at x
-        :rtype: float
+        :returns: Recall at n_recommendations
+        :rtype: numpy.float16
         """
         recalls = []
         for user in range(self.ratings.shape[0]):
             top_recommendations = TopRecommendations(n_recommendations)
-            ctr = 0
-            liked_items = 0
-            for rating in predictions[user, :]:
+            for ctr, rating in enumerate(predictions[user, :]):
                 top_recommendations.insert(ctr, rating)
-                liked_items += rating
-                ctr += 1
             recommendation_hits = 0
             user_likes = self.ratings[user].sum()
-            for index in top_recommendations.get_indices():
+            for index in list(reversed(top_recommendations.get_indices())):
                 recommendation_hits += self.ratings[user][index]
             recall = recommendation_hits / (min(n_recommendations, user_likes) * 1.0)
             recalls.append(recall)
-        return numpy.mean(recalls)
+        return numpy.mean(recalls, dtype=numpy.float16)
+
+    def calculate_ndcg(self, n_recommendations, predictions):
+        """
+        The method calculates the normalized Discounted Cumulative Gain of all users
+        by only looking at the top n_recommendations.
+
+        :param int n_recommendations: number of recommendations to look at, sorted by relevance.
+        :param float[][] predictions: calculated predictions of the recommender
+        :returns: nDCG for n_recommendations
+        :rtype: numpy.float16
+        """
+        ndcgs = []
+        for user in range(self.ratings.shape[0]):
+            top_recommendations = TopRecommendations(n_recommendations)
+            ctr = 0
+            for rating in predictions[user, :]:
+                top_recommendations.insert(ctr, rating)
+                ctr += 1
+            dcg = 0
+            idcg = 0
+            for pos_index, index in enumerate(list(reversed(top_recommendations.get_indices()))):
+                dcg += numpy.power(2, self.ratings[user, index]) - 1 / numpy.log2(pos_index + 2)
+            for pos_index, rating in enumerate(sorted(self.ratings[user, :], reverse=True)):
```

Comment:
The whole idea of calculating top recommendations is to avoid sorting. Here, you can use numpy nonzeros method https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html